### PR TITLE
 integrate Go with optimizer library

### PR DESCRIPTION
**Pull request by dzhwinter**
fix[2516](https://github.com/PaddlePaddle/Paddle/issues/2516)

### CODE REVIEW

File: go/pserver/cclient/test/dump_optimizer_proto.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import OptimizerConfig_pb2 as pb
```
Comment (helinwang) [inline go/pserver/cclient/test/dump_optimizer_proto.py]: What is this file for? If we only need the generated proto file, maybe we can just check in the protofile under `testdata` folder.

File: go/pserver/optimizer.go [hunk 2]
```diff
@@ -1,42 +1,42 @@
...
 	opt *C.struct_paddle_optimizer
 }
 
-func newOptimizer(t optimizerType, learning_rate float64) *optimizer {
+func newOptimizer(paramWithConfigs ParameterWithConfig) *optimizer {
 	o := &optimizer{}
-	o.opt = C.paddle_create_SGD_optimizer(C.double(learning_rate))
+	p := paramWithConfigs.Param
+	c := paramWithConfigs.Config
+	o.opt = C.paddle_create_optimizer(C.uchar(c), C.int(len(c)), unsafe.Pointer(p.Content), c.int(p.Length), nullPtr, 0)
```
Comment (helinwang) [inline go/pserver/optimizer.go]: `paddle_create_optimizer` takes the ownership of the `param_buffer` argument. So this does not work, because `p.Content` is a Go pointer, which is tracked by Go garbage collector, it will be garbage collected. We need to make a copy here.
For more pointer information please see: https://golang.org/cmd/cgo/#hdr-Passing_pointers

Another question is I don't get what is the benefit of changing the type of `p.Content` from `[]byte` to `*byte`:

- Before when `p.Content` have `[]byte` as type, in `paddle_send_grads` the `cArrayToSlice` function does not make a copy of the gradient buffer. It just use the buffer as underlying storage of the created slice. So changing the type to `*byte` does not improve performance.
- At this line of code, we have to make a copy anyway.

So I don't see any benefit of changing the type of `p.Content` from `[]byte` to `*byte`, it comes with added complexity: need a separate variable p.Length to track length, Go programmers are more familiar with slice (`[]byte`) than byte pointers (`*byte`).

File: go/pserver/service.go [hunk 3]
```diff
@@ -44,14 +45,16 @@ type Gradient Parameter
 type Service struct {
 	initialized chan struct{}
 
-	mu       sync.Mutex
-	opt      *optimizer
+	mu sync.Mutex
+	// injection from parameter to optimizer
+	optMap   map[string]*optimizer
```
Comment (helinwang) [inline go/pserver/service.go]: If we already have `optMap` and Optimizer owns the parameter memory, we no longer need `paramMap`.

File: go/pserver/optimizer.go (hunk 2 above)
Comment (dzhwinter) [inline go/pserver/optimizer.go]: I see. fix the copy done.
>Another question is what is the benefit of changing the type of p.Content from []byte to *byte:
there is no benefit,  c style language prefers pointer type.  fix it

File: go/pserver/service.go (hunk 3 above)
Comment (dzhwinter) [inline go/pserver/service.go]: true. fixed

File: go/pserver/cclient/test/dump_optimizer_proto.py (hunk 1 above)
Comment (dzhwinter) [inline go/pserver/cclient/test/dump_optimizer_proto.py]: fix done.

File: go/pserver/cclient/cclient.go [hunk 4]
```diff
@@ -123,9 +123,8 @@ func paddle_begin_init_params(client C.paddle_pserver_client) C.int {
 func paddle_init_param(client C.paddle_pserver_client, param C.paddle_parameter, param_config unsafe.Pointer, config_len C.int) C.int {
 	et := pserver.ElementType(param.element_type)
 	name := C.GoString(param.name)
-	content := cArrayToSlice(unsafe.Pointer(param.content), int(param.content_len))
 	pc := pserver.ParameterWithConfig{
-		Param:  pserver.Parameter{Name: name, ElementType: et, Content: content},
+		Param:  pserver.Parameter{Name: name, ElementType: et, Content: param.content, Length: para.content_len},
```
Comment (dzhwinter) [inline go/pserver/cclient/cclient.go]: fix done. sorry for the typo mistake, can not pass the go link problem yesterday.

**Comment by dzhwinter** [comment]
it is accidentally closed after a force update when I resolved confict. l had tried many methods, it can not be reopened, weird : (
https://github.com/isaacs/github/issues/361

**Comment by dzhwinter** [comment]
If there is any method that can force update the branch and reopen this PR, Please let me know.
create a related PR here:
https://github.com/PaddlePaddle/Paddle/pull/2610
